### ISSUE TITLE
ERROR: value too long for type character varying(8) (`authentication_log`.`request_method`)

### ISSUE THREAD

**Issue by SMores** [authelia/authelia#3482]
### Description

I’m using Authelia with caddy server 2.5.1, using the new forward_auth directive. I have Authelia hosted at authelia.mydomain.com, and a wiki hosted at wiki.mydomain.com. Logging in to Authelia directly via authelia.mydomain.com works without issue, and once logged in, I’m able to access wiki.mydomain.com.

If I’m not logged in, when I try to access wiki.mydomain.com, I’m correctly redirected to the Authelia login page. All authentication attempts fail at this point, because of a SQL error (I’m using Postgres): `ERROR: value too long for type character varying(8)`.

### Expected Behaviour

Authentication with proper credentials should succeed, even if the redirect query param is set!

### Additional Information

Relevant software:

* Authelia v4.35.5
* Caddy v2.5.1
* Postgres v14.3

Error logs:

```
time="2022-06-04T23:38:55-04:00" level=debug msg="Mark 1FA authentication attempt made by user '<REDACTED>'" method=POST path=/api/firstfactor remote_ip=<REDACTED>
time="2022-06-04T23:38:55-04:00" level=error msg="Unable to mark 1FA authentication attempt by user '<REDACTED>': error inserting authentication attempt for user '<REDACTED>': ERROR: value too long for type character varying(8) (SQLSTATE 22001)" method=POST path=/api/firstfactor remote_ip=<REDACTED>
stack="github.com/authelia/authelia/v4/internal/handlers/response.go:223           markAuthenticationAttempt
github.com/authelia/authelia/v4/internal/handlers/handler_firstfactor.go:68 FirstFactorPOST.func1
github.com/authelia/authelia/v4/internal/middlewares/authelia_context.go:53 AutheliaMiddleware.func1.1
github.com/authelia/authelia/v4/internal/middlewares/headers.go:25          SecurityHeadersCSPNone.func1
github.com/authelia/authelia/v4/internal/middlewares/headers.go:35          SecurityHeadersNoStore.func1
github.com/authelia/authelia/v4/internal/middlewares/headers.go:16          SecurityHeaders.func1
github.com/fasthttp/router@v1.4.9/router.go:414                             (*Router).Handler
github.com/valyala/fasthttp@v1.36.0/http.go:153                             (*Response).StatusCode
github.com/valyala/fasthttp@v1.36.0/server.go:2338                          (*Server).serveConn
github.com/valyala/fasthttp@v1.36.0/workerpool.go:224                       (*workerPool).workerFunc
github.com/valyala/fasthttp@v1.36.0/workerpool.go:196                       (*workerPool).getCh.func1
runtime/asm_amd64.s:1571                                                    goexit"
```

The field that the error message is referring to is the `request_method` field of the `authentication_logs` table. It appears that Authelia is attempting to write a much longer string than just the HTTP method name in this situation. Here’s an example:

`GET, https://authelia.mydomain.com/?rd=https://wiki.mydomain.com/,GET`

For now, I was able to resolve this by altering the column type to be `character varying(150)`. 

Here’s my Caddy configuration, in case it’s relevant (I’m not sure where this “request_method” string is coming from, Authelia or Caddy):

```
authelia.shanemoore.me {
  reverse_proxy <REDACTED>
}

wiki.shanemoore.me {
  forward_auth <REDACTED (Authelia address and port)> {
    uri /api/verify?rd=https://authelia.mydomain.com
    copy_headers Remote-User Remote-Groups Remote-Name Remote-Email
  }
  @localnet {
    remote_ip 127.0.0.1 192.168.0.0/16 172.16.0.0/12 10.0.0.0/8
  }

  reverse_proxy @localnet <REDACTED>
}
```

**Comment by james-d-elliott** [comment]
It's detected when the redirection occurs, for which you've not supplied the logs.

**Comment by stndrs** [comment]
Hi, this issue came up for me as well. So far it's happening on my iOS devices. The requests work fine on my laptop running Fedora 36. The relevant logs:
```
time="2022-06-11T00:51:47-04:00" level=debug msg="Mark 1FA authentication attempt made by user 'REDACTED'"
method=POST path=/api/firstfactor remote_ip=REDACTED
time="2022-06-11T00:51:47-04:00" level=error msg="Unable to mark 1FA authentication attempt by user 'REDACTED':
error inserting authentication attempt for user 'REDACTED':
Error 1406: Data too long for column 'request_method' at row 1" method=POST path=/api/firstfactor remote_ip=REDACTED 
stack="github.com/authelia/authelia/v4/internal/handlers/response.go:223
markAuthenticationAttempt\ngithub.com/authelia/authelia/v4/internal/handlers/handler_firstfactor.go:68
FirstFactorPOST.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/authelia_context.go:53
AutheliaMiddleware.func1.1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:25
SecurityHeadersCSPNone.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:35
SecurityHeadersNoStore.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:16
SecurityHeaders.func1\ngithub.com/fasthttp/router@v1.4.9/router.go:414
(*Router).Handler\ngithub.com/valyala/fasthttp@v1.37.0/http.go:153
(*Response).StatusCode\ngithub.com/authelia/authelia/v4/internal/middlewares/strip_path.go:22
StripPath.func1.1\ngithub.com/valyala/fasthttp@v1.37.0/server.go:2338
(*Server).serveConn\ngithub.com/valyala/fasthttp@v1.37.0/workerpool.go:224
(*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.37.0/workerpool.go:196
(*workerPool).getCh.func1\nruntime/asm_amd64.s:1571 goexit"
```

To see the data that was failing to be inserted, I altered the `authentication_logs` table by making `request_method` `VARCHAR(256)` rather than `VARCHAR(8)`. By altering this column, the authentication requests from my iOS devices finally succeeded.

Normally, successful authentication requests insert `GET` into the `request_method` column. After altering `request_method` as described above, the requests from iOS devices are trying to insert `GET, https://authelia.example.com/?rd=https://redirect.example.com/,GET`

The following screenshot shows the table with data from before and after the alteration was made. I've kept sensitive data out of this screenshot but you can see that the later rows include `https://...`

![image](https://user-images.githubusercontent.com/18623094/173174098-5e4a8390-ea89-4eee-a4c6-fb6a97efcc45.png)

**Comment by james-d-elliott** [comment]
Interesting.. I'm wondering what causes this behavior. What proxy are you using @stndrs?

**Comment by stndrs** [comment]
I'm using caddy as my reverse proxy. On my linux machines the URL is `https://auth.example.com/?rd=https%3A%2F%2Fdash.example.com%2F&rm=GET` and that works fine.
I just took a look at the url on iOS and it is different: `https://auth.example.com/?rd=https%3A%2F%2Fdash.example.com%2F&rm=GET,%20https://auth.example.com/?rd=https://dash.example.com/,GET`

Not sure why the redirect URL is different on iOS devices. 

Tested OS/Browsers:
|OS|Browser|Works?|
|---|---|---|
|Fedora 36|Firefox|✅|
|Ubuntu 22.04|Firefox|✅|
|Ubuntu 22.04|Chrome|✅|
|iOS 15.5|Safari|❌|
|iPad OS 15.5|Safari|❌|

**Comment by james-d-elliott** [comment]
Potentially, I'm just trying to replicate it at this stage. I'm also creating a build which may fix it. I'm using an iPad with iOS 12.5 with Traefik and can't  yet replicate it. It could either be traefik is doing some behind the scenes corrections which Caddy isn't or iOS is doing something weird on newer versions. Super strange it's not appearing on other OS's.

Can you do two things? 

1. Show me the forward_auth portion of your Caddyfile.
2. Try this image: `authelia/authelia:fix-redirect-escape`.

**Comment by stndrs** [comment]
Caddyfile snippet:
```
    @auth host auth.example.com
    handle @auth {
        reverse_proxy authelia:9091
    }

    @foo host foo.example.com
    handle @foo {
        forward_auth authelia:9091 {
            uri /api/verify?rd=https://auth.example.com
            copy_headers Remote-User Remote-Groups Remote-Name Remote-Email
        }
        reverse_proxy foo:80
    }
```
I tried that image and it's still having the same problem. Same error if the table is set back to varchar(8), but working with more space in that column.

**Comment by james-d-elliott** [comment]
Can you try `uri /api/verify?rd=https%3A%2F%2Fauth.example.com`?

**Comment by SMores** [comment]
Sorry for the delay; I had an issue with my GitHub notifications that I just resolved.

Thanks for jumping in, @stndrs! I am also using a iPad, on iPadOS 15.4.1. I’ve only tried accessing Authelia on iOS devices; I didn’t even suspect that might be related!

My Authelia version is 4.35.5. This is an example of what’s being logged in the `request_method` field of the `authentication_logs` table (this whole string, including the two `GET`s:

```
GET, https://authelia.mydomain.com/?rd=https://wiki.mydomain.com/,GET
```

And I’m seeing similar, odd redirect URLs on my iPad:

```
https://authelia.example.com/?rd=https%3A%2F%2Fsomeservice.example.com%2F&rm=GET,%20https://authelia.example.com/?rd=https://some service.example.com/,GET
```

**Comment by james-d-elliott** [comment]
> Can you try `uri /api/verify?rd=https%3A%2F%2Fauth.example.com`? Basically `%3A%2F%2F` instead of `://`.

**Comment by nightah** [comment]
Also out of curiosity has anybody tried Chrome and/or Firefox on iPad OS?

I know definitively that Firefox works on iOS 15.5 as I use it myself.

**Comment by SMores** [comment]
@nightah I’m experiencing this with both Firefox and Safari in iPadOS (using caddy, though).

@james-d-elliott with the encoded URI, I’m getting this redirect URL (which I believe is the same as before):

```
https://authelia.example.com/?rd=https%3A%2F%2Fsomeservice.example.com%2F&rm=GET,%20https://authelia.example.com/?rd=https://some service.example.com/,GET
```

And the `request_method` is still:

```
GET, https://authelia.example.com/?rd=https://someservice.example.com/,GET
```

**Comment by james-d-elliott** [comment]
How peculiar. I'll setup a debug version to see if this issue is originating with Authelia or Caddy. Probably the same image.

**Comment by james-d-elliott** [comment]
Sorry about the delay, I've been swamped with personal stuff and PR's. I'll look at this today.

**Comment by francislavoie** [comment]
I did some re-reading of the Caddy code and I realized there was some unnecessary logic which was the likely cause of doubled-up headers https://github.com/caddyserver/caddy/pull/4847.

If you'd like to test it, you can build from that branch with `xcaddy build forward-auth-fix-double-headers`. If you're running in Docker, you can use the `builder` image variant (see the docs on https://hub.docker.com/_/caddy)

**Comment by nightah** [comment]
I can confirm that definitely is the cause of the issue, I set up a test bed with the Advanced configuration example and excluded the `Location` header first, this resulted in the appropriate behaviour in Authelia.

Then I removed the `copy_response_headers` directive entirely and all the duplicated headers were no longer duplicated.

![image](https://user-images.githubusercontent.com/3339418/174767611-6bdfe99f-a886-48a7-9379-78edb3e6791c.png)

**Comment by james-d-elliott** [comment]
It's kinda funny that a bug like this large exists in Safari in a way which it makes a smaller bug breaking. Glad you guys managed to figure this out, I thought this was going to be a nightmare.

**Comment by james-d-elliott** [comment]
This was also fixed on our end in 095ebae7a0edf8d0b99b94ad4561a8ea706d99bf. If anyone needs help either building caddy with the fix or using the advanced example (we fixed that in the mentioned commit, I'd recommend building the fixed caddy version over the advanced example) then my suggestion is opening a [Q&A Discussion](https://github.com/authelia/authelia/discussions/new?category=q-a).

**Comment by francislavoie** [comment]
Caddy v2.5.2 is released btw!
